### PR TITLE
added mssql provider

### PR DESCRIPTION
**Pull request by rob2244**
Not completely done, still need to spin up a server and try to run some of the queries and mutations and need to write some unit tests, but I figured I'd open it up for review/questions/comments.

I added the mssql provider and added a factory so all of the code generation stuff can get resolved at run time. I also had to change the processedField class to take in a translation functions for data types because the data types are different between sql server and postgres. Additionally I moved the export declaration and connection generation logic to the provider cause the mssql provider needs to be able to pass an already existing connection in, as the connection is instantiated asynchronously

### CODE REVIEW

**Comment by jtabalujan2** [comment]
Can we get on a call later this week and have you sorta walk through some of this with me?

Thanks,
Julian Tabalujan

________________________________

File: src/server/DBMetadata/columnTypeTranslators.js [hunk 1]
```diff
@@ -0,0 +1,71 @@
+function pgSQL(type) {
```
Comment (laueian) [inline src/server/DBMetadata/columnTypeTranslators.js]: Does this column type translator support all pg and ms data types?

File: src/server/DBMetadata/metadataProcessor.js [hunk 2]
```diff
@@ -0,0 +1,77 @@
+const ProcessedField = require("./classes/processedField");
```
Comment (laueian) [inline src/server/DBMetadata/metadataProcessor.js]: Did this file change at all or did you just move it?

File: src/server/DBMetadata/mssql/msMetadataRetriever.js [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+  let connectionString = "";
+  const port = info.port || 1433;
+  connectionString += `mssql://${info.user}:${info.password}@${
+    info.host
+  }:${port}/${info.database}?encrypt=true&request%20timeout=${30000}`;
+  return connectionString;
+}
+
+async function getDbPool(connString) {
+  const hash = crypto.createHash("sha256");
```
Comment (laueian) [inline src/server/DBMetadata/mssql/msMetadataRetriever.js]: This is so best practice I can't even

File: src/server/DBMetadata/classes/processedField.js [hunk 4]
```diff
@@ -1,9 +1,11 @@
+const util = require("../../util");
+
 class ProcessedField {
-  constructor(col, tblIdx, fieldIdx) {
+  constructor(col, tblIdx, fieldIdx, translateColumnType) {
```
Comment (laueian) [inline src/server/DBMetadata/classes/processedField.js]: I like that you made this a dependency and didn't just include the file as it we're a helper library +1

File: src/server/Generators/classes/pgSqlProvider.js [hunk 5]
```diff
@@ -70,6 +70,23 @@ class PgSqlProvider {
 
     return query;
   }
+
+  parameterize() {
```
Comment (laueian) [inline src/server/Generators/classes/pgSqlProvider.js]: I see you include a pg refactor in your mssql PR =P

File: src/server/Generators/classes/typeBuilder.js [hunk 6]
```diff
@@ -58,9 +58,11 @@ class TypeBuilder {
   }
 
   createSubQuery(column, processedMetadata) {
```
Comment (laueian) [inline src/server/Generators/classes/typeBuilder.js]: Great refactor, a lot easier to read +1

File: src/server/dbFactory.js [hunk 7]
```diff
@@ -0,0 +1,36 @@
+const PgSqlProvider = require("./Generators/classes/pgSqlProvider");
```
Comment (laueian) [inline src/server/dbFactory.js]: Can you provide an article referencing the pattern you used here and maybe comment a little on why it you decided to use this pattern pleaseeee

File: src/server/DBMetadata/columnTypeTranslators.js (hunk 1 above)
Comment (rob2244) [reply src/server/DBMetadata/columnTypeTranslators.js]: @laueian The mssql supports all the common mssql value types, but I didn't check the pgsql one. I added checking the pgsql one as a separate ticket.

File: src/server/DBMetadata/metadataProcessor.js (hunk 2 above)
Comment (rob2244) [reply src/server/DBMetadata/metadataProcessor.js]: @laueian I just moved/renamed it

File: src/server/DBMetadata/classes/processedField.js (hunk 4 above)
Comment (rob2244) [reply src/server/DBMetadata/classes/processedField.js]: @laueian easier to switch out translation functions that way

File: src/server/Generators/classes/pgSqlProvider.js (hunk 5 above)
Comment (rob2244) [reply src/server/Generators/classes/pgSqlProvider.js]: @laueian  I had to do it to support the async connection initialization for the mssql library

File: src/server/dbFactory.js (hunk 7 above)
Comment (rob2244) [reply src/server/dbFactory.js]: It's loosely based on the factory design pattern https://www.tutorialspoint.com/design_pattern/factory_pattern.htm

**Pull request merged**